sdk/nodejs/runtime/rpc.ts

```diff
@@ -516,25 +516,28 @@ export function deserializeProperty(prop: any): any {
 
                     const urnParts = urn.split("::");
                     const qualifiedType = urnParts[2];
+                    const urnName = urnParts[3];
 
                     const type = qualifiedType.split("$").pop()!;
                     const typeParts = type.split(":");
-                    let pkgName = typeParts[0];
+                    const pkgName = typeParts[0];
                     const modName = typeParts.length > 1 ? typeParts[1] : "";
                     const typName = typeParts.length > 2 ? typeParts[2] : "";
                     const isProvider = pkgName === "pulumi" && modName === "providers";
+
                     if (isProvider) {
-                        pkgName = typName;
+                        const resourcePackage = resourcePackages.get(packageKey(typName, version || ""));
+                        if (!resourcePackage) {
+                            throw new Error(`Unable to deserialize provider ${urn}, no resource package is registered for ${typName}.`);
+                        }
+                        return resourcePackage.constructProvider(urnName, type, {}, { urn });
                     }
 
-                    const resourcePackage = resourcePackages.get(packageKey(pkgName, version || ""));
-                    if (!resourcePackage) {
-                        throw new Error(`Unable to deserialize resource URN ${urn}, no resource package is registered for type ${type}.`);
+                    const resourceModule = resourceModules.get(moduleKey(modName, version || ""));
+                    if (!resourceModule) {
+                        throw new Error(`Unable to deserialize resource ${urn}, no module is registered for ${modName}.`);
                     }
-                    const urnName = urnParts[3];
-                    return !isProvider ?
-                        resourcePackage.construct(urnName, type, {}, { urn }) :
-                        resourcePackage.constructProvider(urnName, type, {}, { urn });
+                    return resourceModule.construct(urnName, type, {}, { urn });
```

Comment:
Passing the empty `{}` for the args is just point-in-time until `getResource` from your other PR is exposed to get the actual state?

Reply:
Yes, that's right.